README.md
# Stalys

![All Modules](/img/allmodules-hornet.png)

## Abstract

This Plugin will contain the modules that I am creating in my study of [cybernetics](https://en.m.wikipedia.org/wiki/Cybernetics) in music.

My goal is to create a [Turing-complete](https://en.m.wikipedia.org/wiki/Turing_completeness) set of instructions for VCV. I would also use it to teach my child about computing and programming :).

My dream is to one day port it to eurorack (but I'd need to learn electronics first 😅).

## Credits

- Neuron original design: [Andrew Fitch](https://www.nonlinearcircuits.com/)
- Code and Panels: [Stalys](https://linktr.ee/stalysmusic)
- Neuron VCV implementation: [Michael Hetrick](https://mhetrick.com/)
- Befaco slew limiter VCV implementation: [Befaco](https://github.com/VCVRack/Befaco)
- BogAudio skinable module classes: [BogAudio](https://github.com/bogaudio/BogaudioModules)

## Installation

- **Dependencies:** Follow the steps in the [official VCV Rack manual](https://vcvrack.com/manual/Building#setting-up-your-development-environment). You do not need to build the Rack itself.
- **Clone  variablethis repository in the destination of your choice:** `git clone https://github.com/denischevalier/StalysVCVPlugin`.
- **Go into the plugin directory:** `cd StalysVCVPlugin`.
- **Setup the environmnent:** `export RACK_DIR=path/to/your/sdk/installation`.
- **Build and install the modules:** `make install`.

## Examples

Some example and real-life usage of this plugin can be found under the [/patches](/patches) directory.

## Skins

New with v1.4.0 are four different skins for Stalys modules:

<details><summary>Bee</summary>

![All Modules](/img/allmodules-bee.png)

</details>

<details><summary>Hornet</summary>

![All Modules](/img/allmodules-hornet.png)

</details>

<details><summary>Light</summary>

![All Modules](/img/allmodules-light.png)

</details>

<details><summary>Dark</summary>

![All Modules](/img/allmodules-dark.png)

</details>

## Modules

<details><summary>VC Dual Neuron</summary>

![VC Dual Neuron](/img/vcdualneuron.png)

This module is heavily inspired by [Nonlinear Circuits](https://www.nonlinearcircuits.com/) [Double Neuron](https://github.com/mhetrick/nonlinearcircuits) module.

The module is divided in three parts. On the left are two VC Neurons circuits, and on the right the original Combiner circuit.

Each Neuron is comprised of three lanes: One input, that is multiplied (four-quadrant multiplier). After multiplication, a VC offset is applied. The three inputs are summed and run into a half-wave rectifier and comparator. The comparator's bipolar output is summed with the rectified signal. SENSE adds a positive DC signal to the mix, while RESPONSE affects the magnitude of the comparator's output.

The Combiner circuit consists of two lanes (with four-quadrant multiplication and vc offset), normalized to each Neuron's output. The output of the two lanes runs into several logic modules:

- A Difference Rectifier (DIFFRECT), which is another creative mixer: the two lanes are summed. Then the negative mix is substracted from the positive mix. If the result is positive, it goes out the DIFFRECT+ output, otherwise it goes out the DIFFRECT- output.
- A MAX circuit which outputs the max value out of the two lanes.
- A SUM output which adds the two lanes.
- A DIFF output which substracts the bottom lane from the top lane.
- A MIN circuit which outputs the min value out of the two lanes.

This might seem overwhelming, but just experiment: feed it with LFOs, envelopes, sequences, oscillators or even itself and watch the complex waveforms it generates. You can use just some parts of it and it becomes a mixer/ring-modulator/waveshapper/logic module.

</details>

<details><summary>8-bit Shift Register</summary>

![ASR](/img/asr.png)

This module is an 8 step shift register: At each incoming CLOCK trigger, all output voltages are shifted down one position and the value of IN is sampled into the first output.

Typical use would be to generate canon melodies.

</details>

<details><summary>8 Comparator</summary>

![8 Comparator](/img/comparator.png)

This module is what its name indicates: 8 VC comparators: It outputs a gate whenever the input signal is superior to the value signal (defaults to 0).

Plug two different LFOs to see complex rhythms emerge.

</details>

<details><summary>8 Multiplier</summary>

![8 Multiplier](/img/multiplier.png)

This module is 8 four quadrant multipliers or ring-modulators (with attenuversion of the Carrier). A 5V voltage is normaled to the Carrier input. The Carrier knob goes from -2 to 2. Between -1 and 1, it will act like a normal four-quadrant. Outside those limits, the signal will be amplified up to ~2 times the original amplitude. This allows it to act as a gain stage before a distortion like our **8 Clip** module.

If no cable is plugged into an output, it is summed to the next outputs.

</details>

<details><summary>8 Summing mixer</summary>

![8 Sum](/img/sum.png)

This module is 8 three inputs unity mixer.

If no cable is plugged into an output, it is summed to the next outputs.

</details>

<details><summary>8 Substractor</summary>

![8 Diff](/img/diff.png)

This module is 8 three inputs difference computer.

</details>

<details><summary>8 Not operator</summary>

![8 Not](/img/not.png)

This module is 8 boolean Not operators: they invert the gates they receive as inputs.

</details>

<details><summary>8 Absolute</summary>

![8 Abs](/img/abs.png)

This modules inverses the negative portion of any input voltage.

</details>

<details><summary>8 VC Sample&Delay</summary>

![8 VC Sample&Delay](/img/vcsampleanddelay.png)

This module is 8 sample and hold with a twist: it will only take gate into account every STEP pulses.

This can serve for counterpoint melody generation, VC clock division, or any usual Sample&Hold usages.

</details>

<details><summary>DA Converter</summary>

![DAC](/img/dac.png)

This modules is an 8 bit bipolar digital to analog converter. It converts input triggers and gates into bipolar CV.

One example of use for it is in a Rungler context: Connect it's output to the frequency of a VCO, the inverted output to the frequency of another VCO. Send the output of VCO 1 to the Analog Shift Register, then the outs of the ASR to the DAC, and use the square out of the second VCO as clock for the ASR: You've got a pseudo-random melody generator.

</details>

<details><summary>AD Converter</summary>

![ADC](/img/adc.png)

This modules is an 8 bit bipolar analog to digital converter. It converts input signal into 8 gates. Gates 1-7 represent the number 1 to 127, where as the 8th is the sign of the signal.

Pair it with DAC and invert or mute some gates to get a pretty gnarly bitcrusher/distortion/waveshaper. Or pass any signal to it to generate interesting rhythms.

</details>

<details><summary>8 Max/or operator</summary>

![8 Max](/img/max.png)

This module is 8 times three input MAX operator logic module (ie. Analog OR). Inputs are normalized to -10 volts to permit to not use all. If an output is not plugged, its value is maxed with the next row, thus allowing up to a 24 inputs max computation.

</details>

<details><summary>8 Min/and operator</summary>

![8 Min](/img/min.png)

This module is 8 times three input MIN operator logic module (ie. Analog AND). Inputs are normalized to 10 volts to permit to not use all. If an output is not plugged, its value is minimized with the next row, thus allowing up to a 24 inputs min computation.

</details>

<details><summary>8 Positive/negative rectifier</summary>

![8 Positive/Negative Rectifier](/img/posnegrectifier.png)

This module is 8 rectifiers that output the positive portion of a signal to the **+** output, and the negative to the **-** output.

This can serve to apply different treatment to each half of a waveform or as a distortion effect.

</details>

<details><summary>8 Averager</summary>

![8 Avg](/img/avg.png)

This module is an octal three-inputs averager: It computes the average of its connected inputs. If an output is not connected, it will compute add the corresponding inputs to the average of the following output, thus permitting averaging of up to 24 inputs.

</details>

<details><summary>8 Function generators</summary>

![8 Function](/img/function.png)

This module is 8 function generators. Each row takes (from left to right) one trigger input, one signal input, one rise CV and param, one fall CV and param. It then outputs the slew processed signal, a gate during rising phase, a gate during falling phase and an End-Of-Cycle gate. Note that EOC is up when no cycle (as on a Serge DUSG), to permit feedback.

This module is a jack of all trades. Amongst its various use-cases, here are the main ones:

* AD envelope generator
* LFO
* VCO
* Gate Sequencer
* Rudimentary low-pass gate
* Filter
* Clock generator
* Slew limiter

Inspired by [Befaco Slew Limiter](https://github.com/VCVRack/Befaco/blob/v1/src/SlewLimiter.cpp).

</details>

<details><summary>8 A/B Y switches</summary>

![8 A/B Y](/img/aby.png)

This module is an octal 2-in 1-out switch, eg A/B Y switch.

It can be used to switch between chorus and verse sequences for instance.

</details>

<details><summary>8 Offset</summary>

![8 Offset](/img/offset.png)

This module is an octal offset module, generating voltages from -10 to 10V.

</details>

<details><summary>8 Clip</summary>

![8 Clip](/img/clip.png)

This module is an octal soft-clipping distortion.

Negative ratio will hard clip the signal. Positive ratio will soft clip the signal. Extreme values will fold the signal.

</details>

<details><summary>8 Clock to V/Oct</summary>

![8 C2V/O](img/ctv.png)

This modules transforms a clock input into a Volt/Octave output. Base frequency (0) is 2 hertz.

Use this module to sync LFOs to a clock.

Note that this modules needs at least two clock ticks to accurately compute the frequency. Thus, low BPM clocks will take some time before being accurately converted.

</details>

<details><summary>Blank</summary>

![Blank](img/blank.png)

This is just your regular 24hp blank panel, themable.

</details>

## Contribute

You can support me in many ways:

* If you want to support me financially, please consider going to [my Patreon](https://www.patreon.com/stalysmusic).
* You can also listen to [my music](https://linktr.ee/stalysmusic).
* If you use my modules in one of your projects (videos, tracks, education, etc.), please send me [an email](mailto:stalysmusic+vcv@gmail.com) for me to check it out. I'll make sure to reference you here.
* You can [create issues](https://github.com/denischevalier/StalysVCVPlugin/issues), improve the [documentation](https://github.com/denischevalier/StalysVCVPlugin/wiki) or create [pull requests](https://github.com/denischevalier/StalysVCVPlugin/pulls) here.
* If you know electronics and are interested in porting one of my modules, please feel free to [contact me](mailto:stalysmusic+vcv@gmail.com).


CHANGELOG.md
<a name="unreleased"></a>
## [Unreleased]

## [1.4.2] - 2020-12-21
### Bug
- Fix build on windows ([#80](https://github.com/denischevalier/StalysVCVPlugin/issues/80))

### Documentation
- Add example patches ([#77](https://github.com/denischevalier/StalysVCVPlugin/issues/77))

<a name="1.4.1"></a>
## [1.4.1] - 2020-12-14
### Bug
- Revert use of -ffast-math on compile ([#76](https://github.com/denischevalier/StalysVCVPlugin/issues/76))

<a name="1.4.0"></a>
## [1.4.0] - 2020-12-09
### Module
- Add 24hp blank panel ([#73](https://github.com/denischevalier/StalysVCVPlugin/issues/73))
- 8 Multiplier can amplify voltage ([#70](https://github.com/denischevalier/StalysVCVPlugin/issues/70))

### Performance
- Use -ffast-math on compile ([#69](https://github.com/denischevalier/StalysVCVPlugin/issues/69))

### UI
- Add alternative skins ([#67](https://github.com/denischevalier/StalysVCVPlugin/issues/67))
- Redraw all panels ([#66](https://github.com/denischevalier/StalysVCVPlugin/issues/66))

<a name="1.3.1"></a>
## [1.3.1] - 2020-12-03
### Bug
- Fix typo in plugin.json ([#60](https://github.com/denischevalier/StalysVCVPlugin/issues/60))

<a name="1.3.0"></a>
## [1.3.0] - 2020-12-02
### Documentation
- Move pictures in sub-folder ([#55](https://github.com/denischevalier/StalysVCVPlugin/issues/55))
- Use collapsible sections for each module description ([#54](https://github.com/denischevalier/StalysVCVPlugin/issues/54))
- Add install guide

### Module
- 8 Clock to Volt/Octave ([#58](https://github.com/denischevalier/StalysVCVPlugin/issues/58))
- Tweak 8 Clip ratio inputs, clamp output ([#57](https://github.com/denischevalier/StalysVCVPlugin/issues/57))
- Add 8 Clip module ([#56](https://github.com/denischevalier/StalysVCVPlugin/issues/56))
- Create 8 Offset module ([#53](https://github.com/denischevalier/StalysVCVPlugin/issues/53))
- Create 8 A/B Y module ([#52](https://github.com/denischevalier/StalysVCVPlugin/issues/52))
- Create 8 Avg module ([#51](https://github.com/denischevalier/StalysVCVPlugin/issues/51))
- Create 8 Function Generators module ([#32](https://github.com/denischevalier/StalysVCVPlugin/issues/32))
- Merge 8 And and 8 Min ([#31](https://github.com/denischevalier/StalysVCVPlugin/issues/31))
- Create 8 Min module ([#24](https://github.com/denischevalier/StalysVCVPlugin/issues/24))
- Create 8 Positive/Negative Rectifier module ([#23](https://github.com/denischevalier/StalysVCVPlugin/issues/23))
- Create 8 Max Module ([#22](https://github.com/denischevalier/StalysVCVPlugin/issues/22))
- Create 8 And module ([#21](https://github.com/denischevalier/StalysVCVPlugin/issues/21))
- Add Analog-to-Digital Converter ([#20](https://github.com/denischevalier/StalysVCVPlugin/issues/20))
- Add Digital-to-Analog Converter ([#19](https://github.com/denischevalier/StalysVCVPlugin/issues/19))
- Snap steps knob and floor steps CV on VCSampleAndDelay ([#18](https://github.com/denischevalier/StalysVCVPlugin/issues/18))
- Create VC Sample&Delay ([#17](https://github.com/denischevalier/StalysVCVPlugin/issues/17))

<a name="1.2.1"></a>
## [1.2.1] - 2020-11-24
### Performance
- Reimplement clip and exponential functions ([#16](https://github.com/denischevalier/StalysVCVPlugin/issues/16))

<a name="1.2.0"></a>
## [1.2.0] - 2020-11-23
### Documentation
- Update TODO list
- Update TODO list

### Module
- Create ASR module ([#15](https://github.com/denischevalier/StalysVCVPlugin/issues/15))
- Create 8 Abs module ([#14](https://github.com/denischevalier/StalysVCVPlugin/issues/14))
- Create 8 Not module ([#13](https://github.com/denischevalier/StalysVCVPlugin/issues/13))
- Create 8 Diff module ([#12](https://github.com/denischevalier/StalysVCVPlugin/issues/12))
- Create 8 Sum module ([#11](https://github.com/denischevalier/StalysVCVPlugin/issues/11))
- Create 8 Multiplier module ([#10](https://github.com/denischevalier/StalysVCVPlugin/issues/10))

<a name="1.1.1"></a>
## [1.1.1] - 2020-11-22
### Bug
- Fix Sense and Response inputs ([#7](https://github.com/denischevalier/StalysVCVPlugin/issues/7))

### Style
- put common functions in a separate file and header. ([#9](https://github.com/denischevalier/StalysVCVPlugin/issues/9))
- Enforce the use of `const` on immutable variables ([#8](https://github.com/denischevalier/StalysVCVPlugin/issues/8))

<a name="1.1.0"></a>
## [1.1.0] - 2020-11-22
### Modules
- Create 8 Comparator module ([#6](https://github.com/denischevalier/StalysVCVPlugin/issues/6))

<a name="1.0.1"></a>
## [1.0.1] - 2020-11-20
### Bug
- Minor interface tweaks: fix overlaping elements (PR [#2](https://github.com/denischevalier/StalysVCVPlugin/issues/2))

### Documentation
- Add CHANGELOG.md (PR [#4](https://github.com/denischevalier/StalysVCVPlugin/issues/4))
- Fix typo in README.md: s/AxBxC/AxB+C/ (PR [#3](https://github.com/denischevalier/StalysVCVPlugin/issues/3))
- Update README.md with abstract, TODO and Contribute (PR [#1](https://github.com/denischevalier/StalysVCVPlugin/issues/1))

<a name="1.0.0"></a>
## 1.0.0 - 2020-11-19

[Unreleased]: https://github.com/denischevalier/StalysVCVPlugin/compare/1.3.1...HEAD
[1.3.1]: https://github.com/denischevalier/StalysVCVPlugin/compare/1.3.0...1.3.1
[1.3.0]: https://github.com/denischevalier/StalysVCVPlugin/compare/1.2.1...1.3.0
[1.2.1]: https://github.com/denischevalier/StalysVCVPlugin/compare/1.2.0...1.2.1
[1.2.0]: https://github.com/denischevalier/StalysVCVPlugin/compare/1.1.1...1.2.0
[1.1.1]: https://github.com/denischevalier/StalysVCVPlugin/compare/1.1.0...1.1.1
[1.1.0]: https://github.com/denischevalier/StalysVCVPlugin/compare/1.0.1...1.1.0
[1.0.1]: https://github.com/denischevalier/StalysVCVPlugin/compare/1.0.0...1.0.1


src/Max.cpp
#include "Common.hpp"

struct Max : StalysModule {
  enum ParamIds { NUM_PARAMS };
  enum InputIds {
    IN1_INPUT,
    IN2_INPUT,
    IN3_INPUT,
    IN4_INPUT,
    IN5_INPUT,
    IN6_INPUT,
    IN7_INPUT,
    IN8_INPUT,
    IN9_INPUT,
    IN10_INPUT,
    IN11_INPUT,
    IN12_INPUT,
    IN13_INPUT,
    IN14_INPUT,
    IN15_INPUT,
    IN16_INPUT,
    IN17_INPUT,
    IN18_INPUT,
    IN19_INPUT,
    IN20_INPUT,
    IN21_INPUT,
    IN22_INPUT,
    IN23_INPUT,
    IN24_INPUT,
    NUM_INPUTS
  };
  enum OutputIds {
    OUT1_OUTPUT,
    OUT2_OUTPUT,
    OUT3_OUTPUT,
    OUT4_OUTPUT,
    OUT5_OUTPUT,
    OUT6_OUTPUT,
    OUT7_OUTPUT,
    OUT8_OUTPUT,
    NUM_OUTPUTS
  };
  enum LightIds { NUM_LIGHTS };

  Max() { config(NUM_PARAMS, NUM_INPUTS, NUM_OUTPUTS, NUM_LIGHTS); }

  void process(const ProcessArgs &args) override {
    const float ins[24] = {inputs[IN1_INPUT].getNormalVoltage(-10.f),
                           inputs[IN2_INPUT].getNormalVoltage(-10.f),
                           inputs[IN3_INPUT].getNormalVoltage(-10.f),
                           inputs[IN4_INPUT].getNormalVoltage(-10.f),
                           inputs[IN5_INPUT].getNormalVoltage(-10.f),
                           inputs[IN6_INPUT].getNormalVoltage(-10.f),
                           inputs[IN7_INPUT].getNormalVoltage(-10.f),
                           inputs[IN8_INPUT].getNormalVoltage(-10.f),
                           inputs[IN9_INPUT].getNormalVoltage(-10.f),
                           inputs[IN10_INPUT].getNormalVoltage(-10.f),
                           inputs[IN11_INPUT].getNormalVoltage(-10.f),
                           inputs[IN12_INPUT].getNormalVoltage(-10.f),
                           inputs[IN13_INPUT].getNormalVoltage(-10.f),
                           inputs[IN14_INPUT].getNormalVoltage(-10.f),
                           inputs[IN15_INPUT].getNormalVoltage(-10.f),
                           inputs[IN16_INPUT].getNormalVoltage(-10.f),
                           inputs[IN17_INPUT].getNormalVoltage(-10.f),
                           inputs[IN18_INPUT].getNormalVoltage(-10.f),
                           inputs[IN19_INPUT].getNormalVoltage(-10.f),
                           inputs[IN20_INPUT].getNormalVoltage(-10.f),
                           inputs[IN21_INPUT].getNormalVoltage(-10.f),
                           inputs[IN22_INPUT].getNormalVoltage(-10.f),
                           inputs[IN23_INPUT].getNormalVoltage(-10.f),
                           inputs[IN24_INPUT].getNormalVoltage(-10.f)};

    const float outs[8] = {std::max(std::max(ins[0], ins[1]), ins[2]),
                           std::max(std::max(ins[3], ins[4]), ins[5]),
                           std::max(std::max(ins[6], ins[7]), ins[8]),
                           std::max(std::max(ins[9], ins[10]), ins[11]),
                           std::max(std::max(ins[12], ins[13]), ins[14]),
                           std::max(std::max(ins[15], ins[16]), ins[17]),
                           std::max(std::max(ins[18], ins[19]), ins[20]),
                           std::max(std::max(ins[21], ins[22]), ins[23])};

    const float out2 = outputs[OUT1_OUTPUT].isConnected()
                           ? outs[1]
                           : std::max(outs[0], outs[1]);
    const float out3 =
        outputs[OUT2_OUTPUT].isConnected() ? outs[2] : std::max(out2, outs[2]);
    const float out4 =
        outputs[OUT3_OUTPUT].isConnected() ? outs[3] : std::max(out3, outs[3]);
    const float out5 =
        outputs[OUT4_OUTPUT].isConnected() ? outs[4] : std::max(out4, outs[4]);
    const float out6 =
        outputs[OUT5_OUTPUT].isConnected() ? outs[5] : std::max(out5, outs[5]);
    const float out7 =
        outputs[OUT6_OUTPUT].isConnected() ? outs[6] : std::max(out6, outs[6]);
    const float out8 =
        outputs[OUT7_OUTPUT].isConnected() ? outs[7] : std::max(out7, outs[7]);

    outputs[OUT1_OUTPUT].setVoltage(outs[0]);
    outputs[OUT2_OUTPUT].setVoltage(out2);
    outputs[OUT3_OUTPUT].setVoltage(out3);
    outputs[OUT4_OUTPUT].setVoltage(out4);
    outputs[OUT5_OUTPUT].setVoltage(out5);
    outputs[OUT6_OUTPUT].setVoltage(out6);
    outputs[OUT7_OUTPUT].setVoltage(out7);
    outputs[OUT8_OUTPUT].setVoltage(out8);
  }
};

struct MaxWidget : StalysModuleWidget {
  static constexpr int hp = 12;

  MaxWidget(Max *module) {
    setModule(module);
    box.size = Vec(RACK_GRID_WIDTH * hp, RACK_GRID_HEIGHT);
    setPanel(box.size, "Max");

    addInput(createInputCentered<port>(mm2px(Vec(7.62, 10.16)), module,
                                       Max::IN1_INPUT));
    addInput(createInputCentered<port>(mm2px(Vec(22.86, 10.16)), module,
                                       Max::IN2_INPUT));
    addInput(createInputCentered<port>(mm2px(Vec(38.1, 10.16)), module,
                                       Max::IN3_INPUT));
    addInput(createInputCentered<port>(mm2px(Vec(7.62, 25.4)), module,
                                       Max::IN4_INPUT));
    addInput(createInputCentered<port>(mm2px(Vec(22.86, 25.4)), module,
                                       Max::IN5_INPUT));
    addInput(createInputCentered<port>(mm2px(Vec(38.1, 25.4)), module,
                                       Max::IN6_INPUT));
    addInput(createInputCentered<port>(mm2px(Vec(7.62, 40.64)), module,
                                       Max::IN7_INPUT));
    addInput(createInputCentered<port>(mm2px(Vec(22.86, 40.64)), module,
                                       Max::IN8_INPUT));
    addInput(createInputCentered<port>(mm2px(Vec(38.1, 40.64)), module,
                                       Max::IN9_INPUT));
    addInput(createInputCentered<port>(mm2px(Vec(7.62, 55.88)), module,
                                       Max::IN10_INPUT));
    addInput(createInputCentered<port>(mm2px(Vec(22.86, 55.88)), module,
                                       Max::IN11_INPUT));
    addInput(createInputCentered<port>(mm2px(Vec(38.1, 55.88)), module,
                                       Max::IN12_INPUT));
    addInput(createInputCentered<port>(mm2px(Vec(7.62, 71.12)), module,
                                       Max::IN13_INPUT));
    addInput(createInputCentered<port>(mm2px(Vec(22.86, 71.12)), module,
                                       Max::IN14_INPUT));
    addInput(createInputCentered<port>(mm2px(Vec(38.1, 71.12)), module,
                                       Max::IN15_INPUT));
    addInput(createInputCentered<port>(mm2px(Vec(7.62, 86.36)), module,
                                       Max::IN16_INPUT));
    addInput(createInputCentered<port>(mm2px(Vec(22.86, 86.36)), module,
                                       Max::IN17_INPUT));
    addInput(createInputCentered<port>(mm2px(Vec(38.1, 86.36)), module,
                                       Max::IN18_INPUT));
    addInput(createInputCentered<port>(mm2px(Vec(7.62, 101.6)), module,
                                       Max::IN19_INPUT));
    addInput(createInputCentered<port>(mm2px(Vec(22.86, 101.6)), module,
                                       Max::IN20_INPUT));
    addInput(createInputCentered<port>(mm2px(Vec(38.1, 101.6)), module,
                                       Max::IN21_INPUT));
    addInput(createInputCentered<port>(mm2px(Vec(7.922, 116.415)), module,
                                       Max::IN22_INPUT));
    addInput(createInputCentered<port>(mm2px(Vec(22.86, 116.84)), module,
                                       Max::IN23_INPUT));
    addInput(createInputCentered<port>(mm2px(Vec(38.1, 116.84)), module,
                                       Max::IN24_INPUT));

    addOutput(createOutputCentered<port>(mm2px(Vec(53.34, 10.16)), module,
                                         Max::OUT1_OUTPUT));
    addOutput(createOutputCentered<port>(mm2px(Vec(53.34, 25.4)), module,
                                         Max::OUT2_OUTPUT));
    addOutput(createOutputCentered<port>(mm2px(Vec(53.34, 40.64)), module,
                                         Max::OUT3_OUTPUT));
    addOutput(createOutputCentered<port>(mm2px(Vec(53.34, 55.88)), module,
                                         Max::OUT4_OUTPUT));
    addOutput(createOutputCentered<port>(mm2px(Vec(53.34, 71.12)), module,
                                         Max::OUT5_OUTPUT));
    addOutput(createOutputCentered<port>(mm2px(Vec(53.34, 86.36)), module,
                                         Max::OUT6_OUTPUT));
    addOutput(createOutputCentered<port>(mm2px(Vec(53.34, 101.6)), module,
                                         Max::OUT7_OUTPUT));
    addOutput(createOutputCentered<port>(mm2px(Vec(53.34, 116.84)), module,
                                         Max::OUT8_OUTPUT));
  }
};

Model *modelMax = createModel<Max, MaxWidget>("Max");
